### PR TITLE
Add a `phpcs.xml.dist` file for PHP file formatting standards

### PR DESCRIPTION
**Pull request by danielbachhuber**

### CODE REVIEW

**Comment by schlessera** [comment]
What is the purpose of this ruleset? To lint the PHP logic added to this repo specifically (like the maintenance commands), or to lint the entire bundle?
I don't think linting the entire bundle is useful, it will just block commits for no reason. Also, it would mean keeping the package-specific exclusions in merged sync with all packages.
The ruleset should instead point to the `.maintenance` folder only, where the PHP files specific to this repository are. It should then only include the rule exclusions that are needed for that part of the code.

**Comment by danielbachhuber** [comment]
> What is the purpose of this ruleset? To lint the PHP logic added to this repo specifically (like the maintenance commands), or to lint the entire bundle?

@schlessera The former: linting the maintenance commands.

> The ruleset should instead point to the `.maintenance` folder only, where the PHP files specific to this repository are. It should then only include the rule exclusions that are needed for that part of the code.

My bad. I copied the `phpcs.xml.dist` from `wp-cli/wp-cli` because I thought we might want to keep them in sync in some fashion.

To make sure we're on the same page, can you annotate the PR with the changes you'd suggest?

File: phpcs.xml.dist [hunk 1]
```diff
@@ -0,0 +1,120 @@
+<?xml version="1.0"?>
+<ruleset name="WP-CLI">
+	<description>Custom ruleset for WP-CLI</description>
```
Comment (schlessera) [inline phpcs.xml.dist:3]: Each ruleset has a distinct name so we can keep them apart.
```suggestion
<ruleset name="WP-CLI-Dev">
	<description>Custom ruleset for the WP-CLI development environment</description>
```

File: phpcs.xml.dist [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+	<!--
+	#############################################################################
+	COMMAND LINE ARGUMENTS
+	For help understanding this file: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Annotated-ruleset.xml
+	For help using PHPCS: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Usage
+	#############################################################################
+	-->
+
+	<!-- What to scan. -->
+	<file>.</file>
```
Comment (schlessera) [inline phpcs.xml.dist:14]: You'll have to test the exact syntax here, but probably something like this:
```suggestion
	<file>./.maintenance/</file>
```

File: phpcs.xml.dist [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+
+	<!-- Ignoring select files/folders.
+		 https://github.com/squizlabs/PHP_CodeSniffer/wiki/Advanced-Usage#ignoring-files-and-folders -->
+	<exclude-pattern>*/tests/data/*</exclude-pattern>
+
```
Comment (schlessera) [inline phpcs.xml.dist:19]: We don't have any tests specific to this package.
```suggestion
```

File: phpcs.xml.dist [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+	<rule ref="WP_CLI_CS">
+		<!-- The `while` control structure is the only one where assignments in conditions are mostly valid. -->
+		<exclude name="WordPress.CodeAnalysis.AssignmentInCondition.FoundInWhileCondition"/>
+
+		<!-- This is a bug in PHPCS. A fix for which has been pulled already and is expected to be
+			 merged in PHPCS 3.5.0. Once this repos upgrades to that version, this exclusion
+			 should be removed. -->
+		<exclude name="PSR2.Namespaces.NamespaceDeclaration.BlankLineAfter"/>
+	</rule>
```
Comment (schlessera) [inline phpcs.xml.dist:43]: Probably no exceptions needed here:
```suggestion
	<rule ref="WP_CLI_CS"/>
```

File: phpcs.xml.dist [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+
+	<!-- Allow for back-compat conversion comments to be explained in logical shorthand.
+		 See: https://github.com/squizlabs/PHP_CodeSniffer/wiki/Customisable-Sniff-Properties#squizphpcommentedoutcode -->
+	<rule ref="Squiz.PHP.CommentedOutCode">
+		<properties>
+			<property name="maxPercentage" value="45"/>
+		</properties>
+	</rule>
+
```
Comment (schlessera) [inline phpcs.xml.dist:85]: These exceptions should not be needed. We might need different ones, though...
```suggestion
```

File: phpcs.xml.dist [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+	<rule ref="WordPress.Arrays.ArrayDeclarationSpacing.AssociativeArrayFound">
+		<exclude-pattern>*/tests/test-(process|utils)\.php$</exclude-pattern>
+	</rule>
+
+	<!-- This is a procedural stand-alone file that is never loaded in a WordPress context,
+		 so this file does not have to comply with WP naming conventions. -->
+	<rule ref="WordPress.NamingConventions.PrefixAllGlobals">
+		<exclude-pattern>*/utils/get-package-require-from-composer\.php$</exclude-pattern>
+	</rule>
+
```
Comment (schlessera) [inline phpcs.xml.dist:119]: These exceptions should not be needed. We might need different ones, though...
```suggestion
```

**Review by schlessera — CHANGES_REQUESTED** [review]

**Comment by danielbachhuber** [comment]
@schlessera It doesn't look like `composer phpcs` works as expected locally...

```
$ composer phpcs
> run-phpcs-tests
sh: run-phpcs-tests: command not found
Script run-phpcs-tests handling the phpcs event returned with error code 127
```

Thoughts on how I should continue to iterate?

Also, should I avoid cloning all of the package repos [in the job](https://github.com/wp-cli/wp-cli-dev/runs/7865981070?check_suite_focus=true)?

image

**Comment by schlessera** [comment]
`run-phpcs-tests` is part of `wp-cli/wp-cli-tests`, so you'd need to pull that in as a development dependency. Not sure what that will produce given the symlinky madness in this package, you'll have to try and see. The `wp-cli/wp-cli-tests` also provides the main PHPCS ruleset you're extending, so you'd need it either way for PHPCS to work correctly.

Regarding the cloning, I guess it would make sense to have the action go through the entire flow. I'm just surprised that GitHub itself has trouble cloning these publicly accessible packages...?!

**Comment by danielbachhuber** [comment]
@schlessera This is fixed up now.

**Review by schlessera — APPROVED** [review]

**Pull request merged**